### PR TITLE
List added/removed content according to a given RepositoryVersion

### PR DESCRIPTION
**Pull request by fabricio-aguiar**
https://pulp.plan.io/issues/5706
closes #5706

Please be sure you have read our documentation on creating PRs:
https://docs.pulpproject.org/en/3.0/nightly/contributing/pull-request-walkthrough.html

### CODE REVIEW

File: CHANGES/5706.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+List added/removed content according to a given RepositoryVersion
```
Comment (daviddavis) [inline CHANGES/5706.feature]: This is quite ambiguous. What about, "Added an optional parameter base_version to RepositoryVersion add() and removed() methods."

Also, this should probably be in `CHANGE/plugin_api/`.

**Comment by daviddavis** [comment]
For extra bonus points, what about adding some unit tests?

File: pulpcore/app/models/repository.py [hunk 2]
```diff
@@ -449,19 +449,33 @@ def artifacts(self):
...
+            base_version (pulpcore.app.models.RepositoryVersion): an optional repository version
+
         Returns:
             QuerySet: The Content objects that were added by this version.
         """
-        return Content.objects.filter(version_memberships__version_added=self)
+        if not base_version:
+            return Content.objects.filter(version_memberships__version_added=self)
+        versions = RepositoryVersion.objects.filter(
+            number__in=range(base_version.number + 1, self.number + 1), repository=self.repository)
```
Comment (daviddavis) [inline pulpcore/app/models/repository.py]: I thought through this logic a bit and I think there is a problem. Consider the case where you have versions 1, 2, 3, 4 and you call added(1) on version 4. If there's a unit that was added in 2 and removed in 3, it'll be returned which is not correct. 

The only way I can think of off the top of my head is to get the content for version 4 and version 1 and subtract.

**Comment by bmbouter** [comment]
@gmbnomis what do you think about this?

File: pulpcore/app/models/repository.py [hunk 3]
```diff
@@ -449,19 +449,59 @@ def artifacts(self):
         """
         Artifact.objects.filter(content__pk__in=self.content)
 
-    def added(self):
+    def added(self, base_version=None):
```
Comment (gmbnomis) [inline pulpcore/app/models/repository.py]: This looks complicated and may take a long time to run if there are many version between the `base_version` and the current version. I was thinking about:

```

    def _relationships(self):
        return RepositoryContent.objects.filter(
            repository=self.repository, version_added__number__lte=self.number
        ).exclude(version_removed__number__lte=self.number)

    @property
    def content(self):
        return Content.objects.filter(version_memberships__in=self._relationships())


    def added(self, base_version=None):
        if not base_version:
            return Content.objects.filter(version_memberships__version_added=self)

        return Content.objects.filter(
            version_memberships__in=self._relationships()
        ).exclude(
            version_memberships__in=base_version._relationships()
        )

    def removed(self, base_version=None):
        if not base_version:
            return Content.objects.filter(version_memberships__version_removed=self)

        return Content.objects.filter(
            version_memberships__in=base_version._relationships()
        ).exclude(
            version_memberships__in=self._relationships()
        )
```
I hadn't the time to look into the cost of the queries in your implementation and in the one from above.

The queries above may be costly (we need a join with the repository version table to get the version numbers to do the range query on version numbers). However, I don't know how this relates to your queries and we do .content queries all the time anyway...

File: pulpcore/tests/unit/models/test_repository.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+        self.assertTrue(self.pks[0] not in removed_pks)
+        self.assertTrue(self.pks[1] in removed_pks)
+        self.assertTrue(self.pks[2] not in removed_pks)
+        self.assertTrue(self.pks[3] not in removed_pks)
+
+        self.assertTrue(self.pks[0] in added_pks)
+        self.assertTrue(self.pks[1] not in added_pks)
+        self.assertTrue(self.pks[2] in added_pks)
+        self.assertTrue(self.pks[3] in added_pks)
```
Comment (gmbnomis) [inline pulpcore/tests/unit/models/test_repository.py]: I would prefer to have additional comparisons in the test e.g. version 3 against all other versions.

File: pulpcore/tests/unit/models/test_repository.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+        added_pks = version3.added(version0).values_list('pk', flat=True)
+        removed_pks = version3.removed(version0).values_list('pk', flat=True)
+
+        self.assertTrue(self.pks[0] not in removed_pks)
+        self.assertTrue(self.pks[1] in removed_pks)
+        self.assertTrue(self.pks[2] not in removed_pks)
+        self.assertTrue(self.pks[3] not in removed_pks)
+
+        self.assertTrue(self.pks[0] in added_pks)
```
Comment (gmbnomis) [inline pulpcore/tests/unit/models/test_repository.py]: Proposal: It's less explicit, but `self.assertCountEqual(added_pks, compress(self.pks, [1, 0, 1, 1]))` is more concise (using compress from itertools).

File: pulpcore/app/models/repository.py [hunk 6]
```diff
@@ -105,7 +105,7 @@ def new_version(self, base_version=None):
                 # now add any content that's in the base_version but not in version
                 version.add_content(base_version.content.exclude(pk__in=version.content))
 
-            if Task.current:
+            if Task.current():
```
Comment (gmbnomis) [inline pulpcore/app/models/repository.py]: Oh, now I know why I had to work around this in my unit tests. Thank you!

File: pulpcore/tests/unit/models/test_repository.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        self.assertEqual(version2.added().count(), 0)
+        self.assertEqual(version3.added().count(), 1)
+
+        self.assertEqual(version0.removed().count(), 0)
+        self.assertEqual(version1.removed().count(), 0)
+        self.assertEqual(version2.removed().count(), 2)
+        self.assertEqual(version3.removed().count(), 0)
+
+        self.assertEqual(version3.added(version0).count(), 3)
+        self.assertEqual(version3.removed(version0).count(), 1)
```
[the remark this answers is not in the document]
Comment (fabricio-aguiar) [reply pulpcore/tests/unit/models/test_repository.py]: I believe I misunderstood the problem, I specifically designed this solution to show that one element was removed during all the versions.

File: pulpcore/app/models/repository.py (hunk 3 above)
Comment (fabricio-aguiar) [reply pulpcore/app/models/repository.py]: definitely your queries are better, thank you!

File: pulpcore/app/models/repository.py [hunk 8]
```diff
@@ -449,19 +449,48 @@ def artifacts(self):
         """
         Artifact.objects.filter(content__pk__in=self.content)
 
-    def added(self):
+    def _relationships(self):
```
Comment (gmbnomis) [inline pulpcore/app/models/repository.py]: I think we should use `_relationships()` in the `content` property as well.

File: pulpcore/app/models/repository.py [hunk 9]
```diff
@@ -449,19 +449,48 @@ def artifacts(self):
...
         """
+        Returns:
+            QuerySet: The Content objects that were added until this version.
```
Comment (gmbnomis) [inline pulpcore/app/models/repository.py]: Docstring needs to be updated

**Comment by gmbnomis** [comment]
@fabricio-aguiar @daviddavis @bmbouter I have written two different fixes for https://pulp.plan.io/issues/5707 based on this fix for 5706. I would love to get feedback on your view on the pros and cons of these implementations, see https://pulp.plan.io/issues/5707#note-8.

@fabricio-aguiar Thank you for the test rig added by this PR!

File: pulpcore/app/models/repository.py [hunk 10]
```diff
@@ -396,6 +396,17 @@ class Meta:
         get_latest_by = 'number'
         ordering = ('number',)
 
+    def _relationships(self):
```
Comment (daviddavis) [inline pulpcore/app/models/repository.py]: This is the only hangup I have with this PR. `_relationships` sounds generic. Would there be a better name?

**Comment by gmbnomis** [comment]
Should this be cherry picked? While it isn't a bugfix, it will be necessary to fix `remove_duplicates()`. https://github.com/pulp/pulpcore/blob/7c1aee68391d19befed1ad33c30d94857b7d7eab/pulpcore/plugin/repo_version_utils.py#L34

File: pulpcore/app/models/repository.py (hunk 10 above)
Comment (bmbouter) [reply pulpcore/app/models/repository.py]: What about `_find_content`? @gmbnomis and @daviddavis what do you think?

Comment (daviddavis) [reply pulpcore/app/models/repository.py]: Hmm, it's not really Content though. It's RepositoryContent. What about `_repository_content`?

Comment (bmbouter) [reply pulpcore/app/models/repository.py]: > Hmm, it's not really Content though. It's RepositoryContent. What about `_repository_content`?

This sounds good to me.

**Comment by bmbouter** [comment]
I think we do want to cherry pick it because we're planning to fix bugs on top of it in 3.0.z like https://pulp.plan.io/issues/5707 ? @daviddavis wdyt?

**Pull request merged**